HSBM.py

```diff
@@ -0,0 +1,391 @@
+        self.vertex_assignments_ = vertex_assignments
+
+    def fit(self, graph, y=None):
+        graph = import_graph(graph)
+        if not self.loops:
+            graph = remove_loops(graph)
+     
+        if y is None:
+            self._estimate_assignments(graph)
+            y = self.vertex_assignments_
+     
+        _, y = check_X_y(graph, y, multi_output=True)
+        if y.shape[1] == 1:
+            # warnings.warn("predicted number of components is 1")
+            # msg = "no blocks found: predicted number of components is 1"
+            # raise ValueError(msg)
+            pass
+        
+        y = y[:, 1:]
+        n_levels = y.shape[1]
+      
+        self.block_weights_ = np.empty(n_levels, dtype=object)
+        self.block_p_ = np.empty(n_levels, dtype=object)
+        self.p_mat_ = np.empty(n_levels, dtype=object)
+
+        for i in range(n_levels):
+            label_single = y[:, i]
+            _, counts = np.unique(label_single, return_counts=True)
+            self.block_weights_[i] = counts / graph.shape[0]
+            block_vert_inds, block_inds, block_inv = _get_block_indices(label_single)
+            block_p = _calculate_block_p(graph, block_inds, block_vert_inds)
+
+            if not self.directed:
+                block_p = symmetrize(block_p)
+            self.block_p_[i] = block_p
+
+            p_mat = _block_to_full(block_p, block_inv, graph.shape)
+            if not self.loops:
+                p_mat = remove_loops(p_mat)
+            self.p_mat_[i] = p_mat
+
+        return self
+
+    def _embed(self, graph, reembedding):
+        # reembedding = False for 1st-time embed before clustering
+        # reembedding = True for embed during clustering if self.reembed=True
+        n_components = self._embed_components(reembedding=reembedding)
+
+        if self.embed_method == "ASE":
+            embedder = AdjacencySpectralEmbed(
+                n_components=n_components, **self.embed_kws
+            )
+            embed = embedder.fit_transform(graph)
+        elif self.embed_method == "LSE":
+            embedder = LaplacianSpectralEmbed(
+                n_components=n_components, **self.embed_kws
+            )
+            embed = embedder.fit_transform(graph)
+
+        return embed
+
+    def _embed_components(self, reembedding):
+        if reembedding is False:
+            if self.n_components is not None:
+                n_components = self.n_components
+            else:
+                n_components = None
+        else:
+            if self.n_components_reembed is not None:
+                # TODO: check this input
+                # (should it be one int or a list of them?)
+                n_components = self.n_components_reembed
+            else:
+                # TODO: calculate max_dim at current level
+                # should be easier if do clustering lvl by lvl in RC
+                pass
+        return n_components
+
+# TODO: subgraph dissimilarity
```

Comment:
what is planned to be used as a dissimilarity here in place of Minh's T statistic in the original hsbm paper?